### ISSUE TITLE
Variable completion in command adds extra blank

### ISSUE THREAD

**Issue by manfredlotz** [fish-shell/fish-shell#5798]
```
fish version is 3.0.2, system is Fedora 29, 

Linux hogwart 5.0.5-200.fc29.x86_64 #1 SMP Wed Mar 27 20:58:04 
UTC 2019 x86_64 x86_64 x86_64 GNU/Linux

TERM is xterm-256color
```

Define a variable like follows:  `set -x MYDIR  '/usr/share/fish/'`

Now type: `cd $MYD<tab>`.  Then completion takes place but an extra space gets added, i.e. if I want to further narrow down the location I want to cd to I first have to go backwards one char before adding a slash aso.

**Comment by 0ion9** [comment]
This is an intentional optimization for the common case where you don't want to add anything onto the end of the variable, and just want to get on with typing the next option or argument.

The alternative would be to not add this extra blank. That would constitute a statement that the case where you *do* want to add onto the end of the expanded variable is more common than the case where you do not want to add on to the end of the expanded variable.

**Comment by manfredlotz** [comment]
It is much easier to type a SPACE than to type a 'left arrow' key (hope the name is correct) which for most keyboards is farer away for the fingers.

**Comment by manfredlotz** [comment]
I think the completion should notice, that the variable is a path, and if the path does exist it should add a '/' if it is a directory. In this case it doesn't matter if the path is absolute or relative. A / (slash) could be the indication that it is a path name.

**Comment by manfredlotz** [comment]
It also is dependent upon the command I'm typing.  For example, `ls` os `cd` indicate that a variable expansion is more likely a path. Additionally, if the flag of a program indicates it is not a path then the completion could add a SPACE. An example for the latter case is the `useradd` command with option 
`-U  -- unlock user's password` which obviously isn't a path name

**Comment by floam** [comment]
Are we sure this isn't https://github.com/fish-shell/fish-shell/issues/5571? Maybe it is possible that we were treating this case differently but it broke for the same reason and this is a regression? I have not tested older fish versions.

**Comment by floam** [comment]
No, the reporter is using 3.0.2 which I don't think is affected.

**Comment by floam** [comment]
> The alternative would be to not add this extra blank. That would constitute a statement that the case where you do want to add onto the end of the expanded variable is more common than the case where you do not want to add on to the end of the expanded variable.

The optimization in question would be: if the variable when expanded is a valid path from the current working directory, let's not insert the extra blank. It doesn't need to be for all situations.

**Comment by manfredlotz** [comment]
@floam I agree this would be the simplest reasoning and in my opinion a sufficient solution. Thanks.

**Comment by BioBox** [comment]
I really want to fix this, but
1. I'm not the biggest fan of C++
2. The project currently seems to be going through a [massive restructuring](#9512)

So I strongly recommend using very easy-to-type variable names for your directories, so that one can type them without needing autocomplete. This is a satisfying work around for me and I'm going to try my best to stick to it.

**Comment by manfredlotz** [comment]
@BioBox I fully disagree. autocompletion should just work.

If the variable points to an existing directory then the autocompletion should add a trailing `/`.

**Comment by BioBox** [comment]
No I totally agree with you. I'm trying to wait until the rust branch is done before I plow though the source sorry.

**Comment by manfredlotz** [comment]
Sorry, I misunderstood. 

You are right. Makes sense to wait for the rust branch.

**Comment by manfredlotz** [comment]
Is this something which is on the agenda for fish v4?

**Comment by faho** [comment]
The issue is open, which means it is in principle something to do, but 4.0 is in beta so it is too late for that release.

**Comment by manfredlotz** [comment]
Thanks for your quick response. Ok, so I hope it will be dealt with after fish v4.0 is out.  😁

**Comment by manfredlotz** [comment]
Tested and things are looking good. 

Thanks a lot for resolving this issue (much sooner than I had expected 😁)
